### PR TITLE
Uniform mvt

### PR DESCRIPTION
**Pull request by Algunenano**
This PR tries to uniformize the MVT interface in Windshaft, with the intention of having the same output using either Mapnik or pg-mvt renderers. As Mapnik's MVTs are already in use it modifies pg-mvt to match its behaviour.

Main changes:
- pg-mvt: Drops the `layer.option.columns` to request only certain columns from the query. It was unused and not implemented in Mapnik's side.
- pg-mvt: Only requests columns of known types in Mapnik (bool, int, float, text).
- pg-mvt: Now supports `geom_column` properly (before it was only working with `the_geom_webmercator` .
- pg-mvt: Now returns an error when the tile is empty (to match Mapnik behaviour)
- Tests: MVT tests are now ignored (instead of omitted) when st_asmvt isn't supported.
- Tests: I've added a new test suit that compares the output of Mapnik and Postgis given the same query. Any test containing the `known_issue` is ignored as the behaviour between the 2 renderers is different.

This is working fine with postgis trunk and I expect it to fail with our 2.4 branch. I don't intend to merge this until the tests pass with out Postgis release, but please have a look at the PR anyway.

Thanks to @enekid and specially @IagoLast for their help dealing with my JS lack of knowledge.

cc/ @davidmanzanares and @rochoa since we were discussing some of the changes in https://github.com/CartoDB/Windshaft/issues/628

### CODE REVIEW

File: lib/windshaft/renderers/pg-mvt/renderer.js [hunk 1]
```diff
@@ -29,70 +38,123 @@ module.exports = class PostgresVectorRenderer {
...
-                );
-                const query = this._replaceTokens(layer.options.sql, {z, x, y});
-
-                queries.push(this._vectorLayerQuery(layer.id, columns, query, { z, x, y }));
+    _getLayerColumns (layer) {
+        return new Promise((resolve, reject) => {
+            if (layer._metadata_columns) {
+                resolve(layer._metadata_columns);
             }
+            var layer_sql = this._replaceTokens(layer.options.sql, 0, 0, 0);
```
Comment (enekid) [inline lib/windshaft/renderers/pg-mvt/renderer.js]: I would use a `const` and the name of the variable should adopt `javascript` conventiones:

`
const layerSQL = this._replaceTokens(layer.options.sql, 0, 0, 0);
`

File: lib/windshaft/renderers/pg-mvt/renderer.js [hunk 2]
```diff
@@ -29,70 +38,123 @@ module.exports = class PostgresVectorRenderer {
...
+            var limitedQuery = `SELECT * FROM (` + layer_sql + `) __windshaft_mvt_schema LIMIT 0;`;
```
Comment (enekid) [inline lib/windshaft/renderers/pg-mvt/renderer.js]: If you use template literals (\`\`), you can use `${}` instead of the `+`s

`
const limitedQuery = `SELECT * FROM (${layer_sql}) __windshaft_mvt_schema LIMIT 0;`;
`

File: lib/windshaft/renderers/pg-mvt/renderer.js [hunk 3]
```diff
@@ -25,74 +34,127 @@ module.exports = class PostgresVectorRenderer {
...
-            const headers = { 'Content-Type': 'application/x-protobuf' };
-            const stats = timer.getTimes();
+                const headers = { 'Content-Type': 'application/x-protobuf' };
+                const stats = timer.getTimes();
 
-            return callback(null, result[0].st_asmvt, headers, stats);
-        });
+                return callback(null, mvt, headers, stats);
+            });
+        }, err => { return callback(err); })
```
Comment (dgaubert) [inline lib/windshaft/renderers/pg-mvt/renderer.js]: To prevent handling error case in two different places, do as follows:

```
Promise.all([ ... ])
  .then(res => ...)
  .catch(err => callback(err))
```
Any error will be caught by `.catch` either the error came from `.all` or from `.then`

File: lib/windshaft/renderers/pg-mvt/renderer.js [hunk 4]
```diff
@@ -25,74 +34,127 @@ module.exports = class PostgresVectorRenderer {
...
+        .catch(() => { }); // Avoids UnhandledPromiseRejectionWarning
```
Comment (dgaubert) [inline lib/windshaft/renderers/pg-mvt/renderer.js]: If something went wrong in `.all` then you are hiding the underlying error. See the comment above. ^^^

File: lib/windshaft/renderers/pg-mvt/renderer.js [hunk 5]
```diff
@@ -25,74 +34,127 @@ module.exports = class PostgresVectorRenderer {
...
+                    }
+                    return callback(err);
+                }
 
-            if (result.length <= 0 || result[0].st_asmvt === undefined) {
-                return callback(new Error(`Couldn't generate tile`));
-            }
+                const mvt = extractMVT(data);
+                if (!mvt) {
+                    return callback(new Error(empty_tile_msg));
```
Comment (dgaubert) [inline lib/windshaft/renderers/pg-mvt/renderer.js]: I understand you are trying to unify the behaviour of both `mapnik` and `pg_asmvt` renderers. `Tilelive bridge` is the one that returns the very same error when the MVT is empty for a given set of coordinates then in `Windshaft-cartodb` we respond with `204 No Content`, among other things, to be able to cache that response at `CDN` level. Thanks to that I realized that we are doing that almost right. The fact of returning an error here makes that we are skipping the generation of cache headers (surrogate headers, cache-channel, etc..). 

So we have to options either modify `tilelive-bridge` to not return an error when the MVT is empty and, also implement the same here or handle this error in [`windshaft-cartodb`](https://github.com/CartoDB/Windshaft-cartodb/blob/028c3f9aeca6f39ad8736fc7d46724b51410d2ab/lib/cartodb/api/map/tile-layergroup-controller.js#L108-L110) to continue the normal req/res cycle generating the headers properly.

I leave to you that decission, we will have to tackle this sooner than later. If you consider that's out of the scope of thid PR, please open new issue.

Comment (Algunenano) [reply lib/windshaft/renderers/pg-mvt/renderer.js]: I think the proper solution is to return a `200` with an empty tile mainly because that removes the need of parsing the error message and the special case. Since it involves modifying other modules I'll create a ticket here to adapt both renderers.

Comment (Algunenano) [reply lib/windshaft/renderers/pg-mvt/renderer.js]: Done in https://github.com/CartoDB/Windshaft/issues/633. I'll tackle it in the near future

Comment (dgaubert) [reply lib/windshaft/renderers/pg-mvt/renderer.js]: thanks!

File: test/acceptance/mvt.js [hunk 6]
```diff
@@ -481,61 +372,604 @@ function mvtTest(usePostGIS) {
...
+            const testClientPg_mvt = new TestClient(mapConfig, { mvt: { usePostGIS: true } });
+            const options = { format : 'mvt' };
+            const z = test.tile && test.tile.z ? test.tile.z : 0;
+            const x = test.tile && test.tile.x ? test.tile.x : 0;
+            const y = test.tile && test.tile.y ? test.tile.y : 0;
+
+            testClientMapnik.getTile(z, x, y, options, function (err1, mapnikMVT) {
+                testClientPg_mvt.getTile(z, x, y, options, function (err2, pgMVT) {
+                    if (err1 || err2) {
+                        assert.ok(err1, "Mapnik didn't fail. Postgis error: " + err1);
```
Comment (dgaubert) [inline test/acceptance/mvt.js]: What happens if we receive an error here but is not the error what we expect? We should separate the tests that check that everything went ok from the tests that something went wrong, and assert the expectations accordingly.

File: test/acceptance/mvt.js [hunk 7]
```diff
@@ -481,61 +372,604 @@ function mvtTest(usePostGIS) {
...
+        it(test.name, function (done) {
+
+            const testClientMapnik = new TestClient(test.mapConfig, { mvt: { usePostGIS: false } });
+            const testClientPg_mvt = new TestClient(test.mapConfig, { mvt: { usePostGIS: true } });
+            const options = { format : 'mvt' };
+
+            testClientMapnik.getTile(test.tile.z, test.tile.x, test.tile.y, options, function (err1, mapnikMVT) {
+                testClientPg_mvt.getTile(test.tile.z, test.tile.x, test.tile.y, options, function (err2, pgMVT) {
+                    if (err1 || err2) {
+                        assert.ok(err1);
```
Comment (dgaubert) [inline test/acceptance/mvt.js]: Shame here. We should separate error cases from the others.

File: lib/windshaft/renderers/pg-mvt/renderer.js [hunk 8]
```diff
@@ -25,74 +34,127 @@ module.exports = class PostgresVectorRenderer {
...
+            const limitedQuery = `SELECT * FROM (${layerSQL}) __windshaft_mvt_schema LIMIT 0;`;
+
+            this.psql.query(limitedQuery, function (err, data) {
+                if (err) {
+                    if (err.message) {
+                        err.message = "PgMvtRenderer: " + err.message;
+                    }
+                    return reject(err);
+                }
+                else {
```
Comment (dgaubert) [inline lib/windshaft/renderers/pg-mvt/renderer.js]: Unnecessary `else`.

File: test/acceptance/mvt.js (hunk 6 above)
Comment (Algunenano) [reply test/acceptance/mvt.js]: Maybe using the `expected_error` more extensively? I mean, declaring it in all the tests that are expected to fail as with the `Null geometry` test case:
```
        {
            name: 'Null geometry',
            sql: "SELECT 2 AS cartodb_id, null as the_geom",
            expected_error : "Error: Tile does not exist"
        },
```

Or would you rather have them all separated by expected to fail and expected to be ok?

Comment (dgaubert) [reply test/acceptance/mvt.js]: Both suggestions are ok. It might that having a separated error case test suite will improve the readability. Don't you think?

Comment (Algunenano) [reply test/acceptance/mvt.js]: I changed to check for the error message for now but, as all of them are errors due to an empty tile and that's gonna be changed, I'll revisit it when that's changed.

File: lib/windshaft/renderers/pg-mvt/renderer.js [hunk 9]
```diff
@@ -25,74 +34,124 @@ module.exports = class PostgresVectorRenderer {
...
+                    const query = this._replaceTokens(layer.options.sql, { z, x, y });
+                    return this._vectorLayerQuery(layer, columns, query, { z, x, y });
+                });
+        });
     }
 
-    _geomColumn ({ z, x, y }) {
+    _geomColumn (layer, { z, x, y }) {
         // Geometry column
         const subpixelBufferSize = Math.round(this.buffer_size * this.mvt_extent / this.tile_size);
```
Comment (oleurud) [inline lib/windshaft/renderers/pg-mvt/renderer.js]: Only a comment. It is not related to your changes, but we could calculate `subpixelBufferSize` once in the constructor.

Comment (Algunenano) [reply lib/windshaft/renderers/pg-mvt/renderer.js]: Nice catch. I also see now that I'm not caching the column names although I check if it's cached (:S). I'll fix that too.

File: lib/windshaft/renderers/pg-mvt/renderer.js [hunk 10]
```diff
@@ -25,74 +34,125 @@ module.exports = class PostgresVectorRenderer {
...
-                const geomColumn = this._geomColumn({z, x, y});
-                const columns = removeDuplicates(
-                    [ geomColumn ].concat(LayerColumns.getColumns(layer.options))
-                );
-                const query = this._replaceTokens(layer.options.sql, {z, x, y});
-
-                queries.push(this._vectorLayerQuery(layer.id, columns, query, { z, x, y }));
+    _getLayerColumns (layer) {
+        return new Promise((resolve, reject) => {
+            if (layer._mvtColumns) {
```
Comment (rochoa) [inline lib/windshaft/renderers/pg-mvt/renderer.js]: This smells. If this is we do just once, it should be in the factory, populated once and injected into the Renderer.

**Comment by Algunenano** [comment]
This is green now with the new 2.4.4.2+carto-1 Postgis release (Thanks @oleurud and @enekid). Do any of you want to have a quick review after all the changes?

The change to not throw an error with empty MVT will come in a different PR.

File: lib/windshaft/backends/tile.js [hunk 11]
```diff
@@ -22,43 +18,41 @@ module.exports = TileBackend;
...
+        if (is_cached) {
+            extraHeaders['X-Windshaft-Cache'] = Date.now() - renderer.ctime;
+        }
 
-    var renderer;
+        timer.start('render-' + params.format.replace('.','-'));
 
-    timer.start('getTileOrGrid');
-    var extraHeaders = {};
-    step(
```
Comment (oleurud) [inline lib/windshaft/backends/tile.js]: :muscle:

File: lib/windshaft/backends/tile.js [hunk 12]
```diff
@@ -22,43 +18,41 @@ module.exports = TileBackend;
 // * cache_buster - specify to ensure a new renderer is used
 // * geom_type - specify default style to use if no style present
 TileBackend.prototype.getTile = function(mapConfigProvider, params, callback) {
+    if (params.format === 'grid.json' && !params.interactivity) {
+        if ( ! params.token ) { // token embeds interactivity
+            callback(new Error("Missing interactivity parameter"));
```
Comment (enekid) [inline lib/windshaft/backends/tile.js]: aren't you missing here a `return`?
```javascript
return callback(new Error("Missing interactivity parameter"));
```

File: lib/windshaft/renderers/pg-mvt/factory.js [hunk 13]
```diff
@@ -20,6 +21,74 @@ function PgMvtFactory(options) {
...
+                700,    // float4
+                701,    // float8
+                1700,   // numeric
+                1042,   // bpchar
+                1043,   // varchar
+                25,     // text
+                705,    // literal
+            ];
+
+            let metadataColumns = [];
```
Comment (enekid) [inline lib/windshaft/renderers/pg-mvt/factory.js]: better use `const`:
```javascript
const metadataColumns = [];
```

File: lib/windshaft/renderers/pg-mvt/renderer.js [hunk 14]
```diff
@@ -54,48 +66,40 @@ module.exports = class PostgresVectorRenderer {
...
-                    [ geomColumn ].concat(LayerColumns.getColumns(layer.options))
-                );
-                const query = this._replaceTokens(layer.options.sql, {z, x, y});
-
-                queries.push(this._vectorLayerQuery(layer.id, columns, query, { z, x, y }));
-            }
-
-            return queries;
-        }, []);
+        return this.layers.filter(l => shouldUseLayer(l, z)).map(layer => {
```
Comment (enekid) [inline lib/windshaft/renderers/pg-mvt/renderer.js]: use `layer`  instead of `l` and IMO spread it across multiple lines
```javascript
return this.layers
                  .filter(layer => shouldUseLayer(layer, z))
                  .map(layer => {
```

File: lib/windshaft/backends/tile.js (hunk 12 above)
Comment (Algunenano) [reply lib/windshaft/backends/tile.js]: You are absolutely right

File: lib/windshaft/renderers/pg-mvt/factory.js (hunk 13 above)
Comment (Algunenano) [reply lib/windshaft/renderers/pg-mvt/factory.js]: I thought `const` was equivalent the c++ const (does not change), but I see it means `won’t be reassigned`. I'll change it.
